### PR TITLE
try use file share and access to implement reader and writer lock

### PR DESCRIPTION
**Pull request by OsmondJiang**
refine reader and writer lock using file access and file share

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5108)

### CODE REVIEW

**Comment by OsmondJiang** [comment]
> wrap `ReleaseMutex` in `finally` block

Just did :)

File: src/docfx/lib/process/InterProcessReaderWriterLock.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            if (_fileStream != null)
+            {
+                using (new GlobalMutex(_filePath))
+                {
+                    _fileStream.Dispose();
+                }
+            }
+        }
+
+        private class GlobalMutex : IDisposable
```
Comment (yufeih) [inline src/docfx/lib/process/InterProcessReaderWriterLock.cs]:
```suggestion
        private struct GlobalMutex : IDisposable
```

File: src/docfx/lib/process/InterProcessReaderWriterLock.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            var filePath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName));
+            var fileLock = WaitFile(lockName, filePath, FileAccess.Write, FileShare.None);
+
+            return new InterProcessReaderWriterLock(fileLock, filePath);
+        }
+
+        private static FileStream WaitFile(string name, string path, FileAccess access, FileShare fileShare)
+        {
+            var start = DateTime.UtcNow;
+            while (DateTime.UtcNow - start <= TimeSpan.FromMinutes(1))
```
Comment (yufeih) [inline src/docfx/lib/process/InterProcessReaderWriterLock.cs]: suggest to repeat forever until we grab the lock, similar to `InterProcessMutex`

File: src/docfx/lib/process/InterProcessReaderWriterLock.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            var fileLock = WaitFile(lockName, filePath, FileAccess.Read, FileShare.Read);
+
+            return new InterProcessReaderWriterLock(fileLock, filePath);
+        }
+
+        public static InterProcessReaderWriterLock CreateWriterLock(string lockName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(lockName));
+
+            var filePath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName));
```
Comment (live1206) [inline src/docfx/lib/process/InterProcessReaderWriterLock.cs]: There is already a hash operation in `GlobalMutex`, do we need a nested hash?

Comment (OsmondJiang) [reply src/docfx/lib/process/InterProcessReaderWriterLock.cs]: image

Comment (OsmondJiang) [reply src/docfx/lib/process/InterProcessReaderWriterLock.cs]: lock name can be remote url

Comment (yufeih) [reply src/docfx/lib/process/InterProcessReaderWriterLock.cs]: `HashUtility.GetMd5Hash(lockName)` can be shared though

**Pull request merged**